core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -434,53 +434,57 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
             implicit transid: TransactionId): Future[(ActivationId, Option[WhiskActivation])] = {
         // merge package parameters with action (action parameters supersede), then merge in payload
         val args = { env map { _ ++ action.parameters } getOrElse action.parameters } merge payload
-        val message = Message(transid, s"/actions/invoke/${action.namespace}/${action.name}/${action.rev}", user, ActivationId(), args)
 
+        val promise = Promise[WhiskActivation]
+        val duration = action.limits.timeout()
+        val timeout = duration + blockingInvokeGrace
+        val activationId = if (blocking) {
+            info(this, s"[POST] action activation will block on result up to $timeout ($duration + $blockingInvokeGrace grace)")
+            // Start listening for finished activations before sending the action to invoker
+            // This avoids, that the invoker is faster with the finished message than the controller knows that it has to listen to it.
+            // If the post of LB to Kafka fails, the activationId will be deleted after 30 seconds from the map, the controller waits for.
+            // Letting the promise fail on a failure on posting to Kafka is not necessary, because nothing will done with it afterwards.
+            pollLocalForResult(user.namespace, ActivationId(), promise)
```

Comment:
As noted in personal discussion:
I found it hard to grasp the code because we use `Unit` returning functions which resolve a `Promise` at some point. This reads like it blocks the code until the result is there (which makes no sense at this point).

I'ld prefer to use `Future` composition if possible so it becomes more apparent that this is in fact returning a Future which is resolved as soon as the ack-message is there.